app-ng/src/server.lisp

```diff
@@ -0,0 +1,112 @@
+;;;; server.lisp
+;;;;
+;;;; Author: Robert Smith
+
+(in-package #:qvm-app-ng)
+
+(alexandria:define-constant +default-server-address+ "0.0.0.0"
+  :test #'string=
+  :documentation "The default host address on which the HTTP server will listen.")
+
+(alexandria:define-constant +default-server-port+ 5000
+  :documentation "The default port on which the HTTP server will listen.")
+
+(defclass rpc-acceptor (tbnl:easy-acceptor)
+  ()
+  (:default-initargs
+   :address (error "Host address must be specified.")
+   :document-root nil
+   :error-template-directory nil
+   :persistent-connections-p t))
+
+(defvar *rpc-acceptor* nil)
+
+(defun start-server-mode (&key host port)
+  (check-type host string)
+  ;; A PORT of 0 tells hunchentoot to pick a random port.
+  (check-type port (or null (integer 0 65535)) "The port must be between 0 and 65535.")
+  (format-log "Starting server on port ~D." port)
+  (start-server host port)
+  (loop (sleep 60)))
+
+(defun start-server (host port &optional debug)
+  (setf tbnl:*log-lisp-backtraces-p* debug
+        tbnl:*log-lisp-errors-p* debug
+        tbnl:*show-lisp-errors-p* debug
+        tbnl:*show-lisp-backtraces-p* debug
+        tbnl:*catch-errors-p* (not debug))
+  (setf *rpc-acceptor* (make-instance
+                        'rpc-acceptor
+                        :address host
+                        :port port
+                        :taskmaster (make-instance 'tbnl:one-thread-per-connection-taskmaster)))
+  (pushnew #'dispatch-rpc-handlers tbnl:*dispatch-table*)
+  (tbnl:reset-session-secret)
+  (tbnl:start *rpc-acceptor*))
+
+(defun stop-server (&optional (acceptor *rpc-acceptor*))
+  (tbnl:stop acceptor))
+
+(defun session-info ()
+  "Return a string describing the server session info."
+  (if (or (not (boundp 'tbnl:*session*))
+          (null tbnl:*session*))
+      ""
+      (format nil
+              "[~A Session:~D] "
+              (tbnl:session-remote-addr tbnl:*session*)
+              (tbnl:session-id tbnl:*session*))))
+
+(defun encode-json (object)
+  (when (boundp 'tbnl:*reply*)
+    (setf (tbnl:content-type*) "application/json; charset=utf-8"))
```

Comment:
How is this going to work with the `amplitudes` (or whatever becomes of it) which streams binary data?

Reply:
You wouldn't use `encode-json`.  I mean you could base64 encode the result or whatever, but it seems crazy to do that for such a (potentially) large payload.

Streaming the bytes back as an `application/octet-stream` like the current approach seems reasonable.

If we go the route of implementing something like rigettizach's suggestion, then presumably you return a plist with some structure that describes what action you want: i.e `:encode-json`, `:stream-octets`, or maybe just let's you specify the content-type for your reply and the thing responsible for actually sending the response figures out what do based on that.

For streaming octets, I guess you could either pass the data and let the response generator handle the streaming, or else provide a callback that accepts a `stream` argument that it can write to.

Reply:
Actually, I'd love to use a standard RPC format like JSON RPC 2.0, but this requirement to stream raw binary back doesn't seem compatible with the spec as far as I can tell. I suppose we could still use JSON RPC for all inputs and "break the law" by sometimes returning a byte stream rather than a JSON RPC response, but that might be more confusing than a 100% "home made" protocol.

Or perhaps with the new job API, we could have every endpoint return a valid JSON RPC response, which might contain a URL where you can go to fetch the bytes for your successfully completed request...

Reply:
we might consider only supporting other means of getting the wf, maybe.................

Reply:
Handlers now return an actual `response` object, rather than calling `encode-json` directly and returning the JSON string. Currently the only two subtypes are `json-response` and `text-response`, but it should be straightforward (at least that was the intention) to add a `streamed-response` or some such for handlers that need it.